File: app/src/main/java/tech/pauly/findapet/shared/MainActivity.java
package tech.pauly.findapet.shared;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import javax.inject.Inject;

import dagger.android.AndroidInjection;
import io.reactivex.disposables.CompositeDisposable;
import tech.pauly.findapet.R;
import tech.pauly.findapet.databinding.ActivityMainBinding;
import tech.pauly.findapet.discover.FilterActivity;
import tech.pauly.findapet.shared.datastore.FilterAnimalTypeUseCase;
import tech.pauly.findapet.shared.datastore.TransientDataStore;
import tech.pauly.findapet.shared.events.ActivityEvent;
import tech.pauly.findapet.shared.events.ViewEventBus;

public class MainActivity extends BaseActivity {

    @Inject
    MainViewModel viewModel;

    @Inject
    ViewEventBus eventBus;

    @Inject
    TransientDataStore dataStore;

    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        AndroidInjection.inject(this);
        super.onCreate(savedInstanceState);
        this.binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        ActivityMainBinding binding = this.binding;
        getLifecycle().addObserver(viewModel);
        binding.setViewModel(viewModel);

        setupDrawer();
    }

    @Override
    public void onResume() {
        super.onResume();
        subscribeToDrawerChange();
        subscribeToExpandingLayoutChange();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_search, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.menu_search:
                dataStore.save(new FilterAnimalTypeUseCase(viewModel.currentAnimalType.get()));
                activityEvent(ActivityEvent.build(this).startActivity(FilterActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Nullable
    @Override
    protected CompositeDisposable registerViewEvents() {
        CompositeDisposable viewEvents = new CompositeDisposable();

        viewEvents.add(eventBus.fragment(MainViewModel.class).subscribe(this::fragmentEvent));

        return viewEvents;
    }

    private void subscribeToDrawerChange() {
        subscribeOnLifecycle(viewModel.getDrawerSubject()
                                      .subscribe(bool -> binding.drawerLayout.closeDrawers(),
                                                 Throwable::printStackTrace));
    }

    private void subscribeToExpandingLayoutChange() {
        subscribeOnLifecycle(viewModel.getExpandingLayoutSubject().subscribe(event -> {
            if (event == MainViewModel.ExpandingLayoutEvent.TOGGLE) {
                binding.expandingLayout.tapToggle();
            } else {
                binding.expandingLayout.collapse();
            }
        }, Throwable::printStackTrace));
    }

    private void setupDrawer() {
        DrawerLayout drawerLayout = binding.drawerLayout;
        ActionBarDrawerToggle drawerToggle;
        Toolbar toolbar = binding.toolbar;
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);

            drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.drawer, R.string.drawer_content);
            drawerToggle.setDrawerIndicatorEnabled(true);

            drawerLayout.addDrawerListener(drawerToggle);
            drawerToggle.syncState();
        }
    }
}

File: app/src/main/java/tech/pauly/findapet/shared/MainActivity.kt
package tech.pauly.findapet.shared

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.ActionBarDrawerToggle
import android.view.Menu
import android.view.MenuItem
import com.crashlytics.android.Crashlytics
import dagger.android.AndroidInjection
import io.fabric.sdk.android.Fabric
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import kotlinx.android.synthetic.main.activity_main.*
import tech.pauly.findapet.R
import tech.pauly.findapet.databinding.ActivityMainBinding
import tech.pauly.findapet.discover.AnimalListItemViewModel
import tech.pauly.findapet.discover.DiscoverViewModel
import tech.pauly.findapet.discover.FilterActivity
import tech.pauly.findapet.favorites.FavoritesViewModel
import tech.pauly.findapet.settings.SettingsViewModel
import tech.pauly.findapet.shared.datastore.FilterAnimalTypeUseCase
import tech.pauly.findapet.shared.datastore.TransientDataStore
import tech.pauly.findapet.shared.events.ActivityEvent
import tech.pauly.findapet.shared.events.OptionsMenuState
import tech.pauly.findapet.shared.events.ViewEventBus
import tech.pauly.findapet.shelters.SheltersViewModel
import javax.inject.Inject

class MainActivity : BaseActivity() {

    @Inject
    internal lateinit var viewModel: MainViewModel

    @Inject
    internal lateinit var eventBus: ViewEventBus

    @Inject
    internal lateinit var dataStore: TransientDataStore

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this)
        super.onCreate(savedInstanceState)

        Fabric.with(this, Crashlytics())

        val binding = DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)
        lifecycle.addObserver(viewModel)
        binding.viewModel = viewModel

        setupDrawer()
    }

    override fun onResume() {
        super.onResume()
        subscribeToDrawerChange()
        subscribeToExpandingLayoutChange()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        return when (id) {
            R.id.menu_search -> {
                viewModel.currentAnimalType.get()?.let {
                    dataStore += FilterAnimalTypeUseCase(it)
                }
                activityEvent(ActivityEvent(this, FilterActivity::class, false))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override val viewEvents: CompositeDisposable?
        get() = CompositeDisposable().also {
            it += eventBus.fragment(MainViewModel::class).subscribe(this::fragmentEvent)
            it += eventBus.optionsMenu(DiscoverViewModel::class).subscribe(this::optionsMenuEvent)
            it += eventBus.optionsMenu(SheltersViewModel::class).subscribe(this::optionsMenuEvent)
            it += eventBus.optionsMenu(FavoritesViewModel::class).subscribe(this::optionsMenuEvent)
            it += eventBus.optionsMenu(SettingsViewModel::class).subscribe(this::optionsMenuEvent)
            it += eventBus.dialog(AnimalListItemViewModel::class).subscribe(this::dialogEvent)
            it += eventBus.dialog(FavoritesViewModel::class).subscribe(this::dialogEvent)
        }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        when (currentMenuState) {
            OptionsMenuState.DISCOVER -> menuInflater.inflate(R.menu.menu_search, menu)
            OptionsMenuState.EMPTY -> menu.clear()
            else -> throw IllegalStateException("OptionsMenuState $currentMenuState not supported in MainActivity")
        }
        return super.onPrepareOptionsMenu(menu)
    }

    private fun subscribeToDrawerChange() {
        viewModel.drawerSubject
                .subscribe({ drawer_layout.closeDrawers() }, Throwable::printStackTrace)
                .onLifecycle()
    }

    private fun subscribeToExpandingLayoutChange() {
        viewModel.expandingLayoutSubject.subscribe({ event ->
            when (event) {
                MainViewModel.ExpandingLayoutEvent.TOGGLE -> expanding_layout.tapToggle()
                else -> expanding_layout.collapse()
            }
        }, Throwable::printStackTrace).onLifecycle()
    }

    private fun setupDrawer() {
        setSupportActionBar(toolbar)
        supportActionBar?.also {
            it.setDisplayHomeAsUpEnabled(true)
            ActionBarDrawerToggle(this, drawer_layout, toolbar, R.string.drawer, R.string.drawer_content).also {
                it.isDrawerIndicatorEnabled = true
                drawer_layout.addDrawerListener(it)
                it.syncState()
            }
        }
    }
}
